Reduce strdup() usage

MrSmile opened the pull request:
I've removed most of strdups from rendering. Error handling should be simpler too.
There is a bunch of strdups in `ass.c` but that's impossible to remove without API change.

There is some concern that low-level parsing functions `mystrto*()` do not respect new string boundaries, but in our case we always have appropriate delimiter after string reference (`argto*()` rely on that too).

TheOneric commented on libass/ass_utils.h:
```diff
@@ -52,6 +52,26 @@ int has_avx(void);
 int has_avx2(void);
 #endif
 
+typedef struct {
+    const char *str;
+    size_t len;
+} ASS_StringView;
```
Several places in the code rely on StringView's `str` to be zero terminated. From the struct alone it may seem like StringView is a fat char pointer — without the need for zero termination. I'd prefer to have a short comment here documenting what StringView is intended to be, iiuc something like this would work:
```C
/* References a substring of some valid C string (0-terminated) */
```

TheOneric reviewed (commented):
Apart from length info leaks, seems good to me.

As you already mentioned, StringView's `str` is passed along as a regular C-string to `mystrto*` and so does `argto` with its args. As `mystro*` is afaict only called via `argto`, which knows the length, and inside `ass_drawing.c: drawing_tokenize`, which also knows the length via StringView, those instances of length info being lost should be straightforward to fix by adding a length parameter to `mystro*`.

Additionally, at least in `ass_fontselect.c: ass_font_select` StringView's `str` is also passed along to several other functions without the length info.
For font strings coming from a Style this won't be a problem since they are zero-terminated right after the font's name.
For font strings originating from a `\fn` string it is not obvious to me why it works, though practical tests suggest that it does in fact work. Those fontnames are substringed from the whole Event text, which usually does not end after the fontname. Afaict it's not until the fontname goes through `font_key_move` *(does it always do this?)*, that a zero terminator is added. In `update_font` it's added to the desc, without any zero-termination.

I'd prefer to have at least a comment in `ass_font_select` explaining why it is safe to discard length info here.

MrSmile commented:
> should be straightforward to fix by adding a length parameter to `mystro*`.

It's not so simple as that requires rewrite of `ass_strtod()` and `strtoll()`. Due to that I decided to defer that change to the future PR.

> Additionally, at least in `ass_fontselect.c`: `ass_font_select` StringView's str is also passed along to several other functions without the length info.

It's intended (I should add corresponding comment there). `ass_font_select()` works with `ASS_Font` copy of cache item key `ASS_FontDesc`. And cache item key is always zero-terminated. Ideally, a cache should have two different key types for request and storage, but that is quite a complication too.

MrSmile commented:
Updated and rebased version. I've reverted changes to `ass_drawing.*` as it's another case where `ASS_StringView` is always zero-terminated, old code works just fine.

astiob commented on libass/ass_cache.c:
```diff
@@ -40,12 +40,13 @@
+    if (!dst)
+        return true;
+
+    ASS_FontDesc *d = dst, *s = dst;
```
I’m confused by these two variables.

MrSmile replied:
Yes, there is a typo here but it's magically not leading to any error.

astiob replied:
Yep 😃 The code works equally well as it is now, and with `*s = src`, and with just a single variable. But either of the last two options would have been easier to read.

astiob commented on libass/ass_cache.c:
```diff
@@ -40,12 +40,13 @@
+        return true;
+
+    ASS_FontDesc *d = dst, *s = dst;
+    memcpy(dst, src, sizeof(ASS_FontDesc));
```
If `s` were to equal `src`, then this could be simplified to `*d = *s;`

MrSmile replied:
I were thinking about that too, but there is a bunch of `memcpy()` in the same context so I decided to use it. Maybe it's indeed better to replace all such `memcpy()` with assignment as that is less error-prone.

astiob commented on libass/ass_cache_template.h:
```diff
@@ -50,6 +50,13 @@
 
 
+START(font, ass_font_desc )
+    STRING(family)
```
`STRING` references `fnv_32a_str`, which you’ve removed.

astiob replied:
No, wait, opposite commit order? GitHub had me.

astiob commented on libass/ass_cache_template.h:
```diff
@@ -50,6 +50,13 @@
+    STRING(family)
+    GENERIC(unsigned, bold)
+    GENERIC(unsigned, italic)
+    GENERIC(int, vertical)  // @font vertical layout
```
(could have made this a `bool` while you were at it)

astiob commented:
Looks good otherwise, and the issues I was too late commenting on are only cosmetic.

Thanks!
